# Optimal Selection of Oil Well Locations

## Table of content

# Introduction

The goal of this project is to determine the optimal sites for the construction of 200 new oil wells by examining three oil exploration zones. We assessed the volume of reserves in each location using a linear regression model, and we used the projected reserves to assess each region's risk and profitability. After evaluating possible risks and returns using bootstrapping techniques, the project suggests exploiting the area with the highest average profit and tolerable risk.

## Objectives

**Selection of the Top 200 Oil Wells per Region**  
The goal is to select the top 200 wells with the highest predicted reserves in each region. This will be done using the predicted values from a linear regression model, ensuring that the wells with the greatest production potential are prioritized for profitability analysis.

**Calculation of Potential Profit from the Selected 200 Wells**  
Once the top 200 wells are selected for each region, the potential profit will be calculated by multiplying the predicted reserves by the unit revenue. This will help assess the financial viability of exploiting the selected wells in each region.

**Risk and Profit Analysis Using Bootstrapping**  
Using the predictions for the top 200 wells, a bootstrapping analysis will be conducted with 1000 samples to estimate the profit distribution. This will include calculating the average profit, 95% confidence interval, and loss risk percentage, helping to identify the region with the best potential for profitable and low-risk oil well exploitation.


## Data Download and Preparation

### Data download

In [2]:
# Data Manipulation
import pandas as pd
import numpy as np

# Machine Learning Models
from sklearn.linear_model import LinearRegression

# Model Evaluation and Metrics
from sklearn.metrics import mean_squared_error
from scipy import stats as st

# Model Selection
from sklearn.model_selection import train_test_split

# Data Resampling
from sklearn.utils import resample

In [5]:
# Load geographical datasets from CSV files
df = pd.read_csv('geo_data_0.csv')
df_1 = pd.read_csv('geo_data_1.csv')
df_2 = pd.read_csv('geo_data_2.csv')

### Data Preparation

#### Region 1

In [6]:
print('General Information')
print()
df.info()  # General information about the dataset
print('='*40)
print('Data Overview')
print()
print(df.head())  # Columns of the dataset
print('='*50)
print('Null Values')
print()
print(df.isnull().sum())  # Count of null values
print('='*40)
print('Duplicate Values')  # Count of duplicate values
df.isna().sum()  # Count of missing values

General Information

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
Data Overview

      id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647
Null Values

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64
Duplicate Values


id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

### Region 2

In [7]:
print('General Information')
print()
df_1.info()  # General information about the dataset
print('='*40)
print('Data Overview')
print()
print(df_1.head())  # Columns of the dataset # CORRECTED BY REVIEWER
print('='*50)
print('Null Values')
print()
print(df_1.isnull().sum())  # Count of null values
print('='*40)
print('Duplicate Values')  # Count of duplicate values
df_1.isna().sum()  # Count of missing values

General Information

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
Data Overview

      id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305
Null Values

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64
Duplicate Values


id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

### Region 3

In [8]:
print('General Information')
print()
df_2.info()  # General information about the dataset
print('='*40)
print('Data Overview')
print()
print(df_2.head())  # Columns of the dataset
print('='*50)
print('Null Values')
print()
print(df_2.isnull().sum())  # Count of null values
print('='*40)
print('Duplicate Values')  # Count of duplicate values
df_2.isna().sum()  # Count of missing values

General Information

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
Data Overview

      id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746
Null Values

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64
Duplicate Values


id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

## Model Training and Testing

### Data Preparation

In [9]:
# Splitting the dataset into features and targets

features = df.drop(['id', 'product'], axis=1)
features_1 = df_1.drop(['id', 'product'], axis=1)
features_2 = df_2.drop(['id', 'product'], axis=1)

targets = df['product']
targets_1 = df_1['product']
targets_2 = df_2['product']

# Splitting the dataset into training and validation sets

features_train, features_valid, targets_train, targets_valid = train_test_split(features, targets, test_size=0.25, random_state=12345)
features_train, features_valid_1, targets_train_1, targets_valid_1 = train_test_split(features_1, targets_1, test_size=0.25, random_state=12345)
features_train, features_valid_2, targets_train_2, targets_valid_2 = train_test_split(features_2, targets_2, test_size=0.25, random_state=12345)

### Model Definition

In [16]:
def Model_Petro(features, targets, features_valid, targets_valid):
    # Initialize and train the linear regression model
    model = LinearRegression()
    model.fit(features, targets)

    # Make predictions on the validation set
    predictions = model.predict(features_valid)

    # Calculate the mean of the predictions
    mean_pred = np.mean(predictions)
    
    # Calculate the RMSE
    rmse = np.sqrt(mean_squared_error(targets_valid, predictions))
    
    # Print the results
    print('The mean of the predictions is:', mean_pred)
    print('The RMSE of the model is:', rmse)
    
    return mean_pred, rmse, predictions

### Region 1

In [17]:
mean_pred, rmse, predictions = Model_Petro(features, targets, features_valid, targets_valid)


The mean of the predictions is: 92.46564781325566
The RMSE of the model is: 37.57547919032473


### Region 2

In [18]:
mean_pred_1, rmse_1, predictions_1 = Model_Petro(features_1, targets_1, features_valid_1, targets_valid_1)


The mean of the predictions is: 68.72718652381815
The RMSE of the model is: 0.8930685055287837


### Region 3

In [19]:
mean_pred_2, rmse_2, predictions_2 = Model_Petro(features_2, targets_2, features_valid_2, targets_valid_2)


The mean of the predictions is: 94.94513027263157
The RMSE of the model is: 40.026749644748875


Based on the results presented so far, it can be said that the region with the best average is region 3. However, region 2 shows the highest precision in its results. Nonetheless, it is still too early to make general judgments for decision-making.

## Profit Calculation

### We define the function

In [22]:
def profitability(mean_pred, invest=100_000_000, wells=200, product_price=4500):
    # Correctly calculate the unit price
    unit_price = invest / (wells * product_price)
    
    # Calculate the threshold
    threshold = invest / (wells * product_price)
    
    if mean_pred >= threshold:
        print('The average amount of reserves will generate profits.')
    else:
        print('The average amount of reserves will result in losses.')

### Calculation by regions

In [23]:
profitability(mean_pred) # Estimations for region 1
profitability(mean_pred_1) # Estimations for region 2
profitability(mean_pred_2) # Estimations for region 3

The average amount of reserves will result in losses.
The average amount of reserves will result in losses.
The average amount of reserves will result in losses.


If we were to randomly select 200 wells in each of the regions, we would incur losses that would make the project unfeasible. Therefore, we need another method to select the best exploitation sites.

## Risks and Profits by Region

In [25]:
def top(predictions, targets):
    data = pd.DataFrame({'predictions': predictions, 'targets': targets})
    top_200 = data.nlargest(200, 'predictions')
    return top_200

def calculate(top):
    revenue = 500000 / 111.1
    total_revenue = top['targets'].sum() * revenue
    investment = 100000000
    profit = total_revenue - investment
    return profit

In [26]:

top_200_1 = top(predictions, targets_valid)
profit_1 = calculate(top_200_1) 

In [27]:
top_200_2 = top(predictions_1, targets_valid_1)
profit_2 = calculate(top_200_2)

In [28]:
top_200_3 = top(predictions_2, targets_valid_2)
profit_3 = calculate(top_200_3)

In [29]:
# Print potential profits
print('Potential results for region 0 USD:', profit_1)
print('Potential results for region 1 USD:', profit_2)
print('Potential results for region 2 USD:', profit_3)

Potential results for region 0 USD: 33221582.589657485
Potential results for region 1 USD: 24163283.295144632
Potential results for region 2 USD: 27336645.095843896


If we select the 200 best wells per region, our results improve significantly. We could have profits, and these could be highly profitable by selecting region 0, which has the best profit margin.

## Calculation of Risks and Profits for Each Region

In [30]:
def top_200(predictions, targets):
    # Convert to numpy arrays
    targets = np.asarray(targets)
    predictions = np.asarray(predictions)
    
    sorted = np.argsort(predictions)[::-1]  # Sort predictions in descending order
    top_200_indices = sorted[:200]  # Get the indices of the top 200 predictions
    selected_predictions = predictions[top_200_indices]  # Select the top 200 predictions
    selected_targets = targets[top_200_indices]  # Select the corresponding targets for the top 200 predictions
    
    return selected_predictions, selected_targets  

def bootstrap_profit(predictions, targets, n_iterations=1000, sample_size=500, investment_cost=100000000, revenue_per_unit=4500):
    state = np.random.RandomState(12345)
    profits = []
    for i in range(n_iterations):
        # Take a bootstrap sample of predictions and targets
        sample_predictions, sample_targets = resample(predictions, targets, n_samples=sample_size, random_state=state)
        
        # Select the top 200 wells from the samples
        _, sample_top_targets = top_200(sample_predictions, sample_targets)
        
        # Calculate the total revenue for this sample
        total_revenue = sample_top_targets.sum() * revenue_per_unit
        profit = total_revenue - investment_cost  # Calculate profit
        profits.append(profit)
    
    # Convert profits to a numpy array
    profits = np.array(profits)
    
    # Calculate the average profit
    avg_profit = profits.mean()
    
    # Calculate the 95% confidence interval
    conf_interval = np.percentile(profits, [2.5, 97.5])
    
    # Calculate the risk of loss (negative profit)
    risk_of_loss = (profits < 0).mean() * 100  # As a percentage
    
    return avg_profit, conf_interval, risk_of_loss

# Using the functions
avg_profit, conf_interval, risk_of_loss = bootstrap_profit(predictions, targets_valid)
avg_profit_1, conf_interval_1, risk_of_loss_1 = bootstrap_profit(predictions_1, targets_valid_1)
avg_profit_2, conf_interval_2, risk_of_loss_2 = bootstrap_profit(predictions_2, targets_valid_2)

In [31]:
# Print the results
print("Region 0: Average profit =", avg_profit, ", 95% CI =", conf_interval, ", Risk of loss =", risk_of_loss, "%")
print("Region 1: Average profit =", avg_profit_1, ", 95% CI =", conf_interval_1, ", Risk of loss =", risk_of_loss_1, "%")
print("Region 2: Average profit =", avg_profit_2, ", 95% CI =", conf_interval_2, ", Risk of loss =", risk_of_loss_2, "%")

Region 0: Average profit = 3964927.666096237 , 95% CI = [-1088816.18926101  9138435.96291906] , Risk of loss = 6.4 %
Region 1: Average profit = 4560451.057866608 , 95% CI = [ 338205.09398985 8522894.53866035] , Risk of loss = 1.5 %
Region 2: Average profit = 4053867.734904668 , 95% CI = [-1680474.36516634  9519024.87781419] , Risk of loss = 7.8 %


Everything seems to indicate that Region 1 has the highest profit margin and a significantly lower percentage of losses, with only 1.5%. However, due to the high risk of 7.8% and the estimated profits being not as significant compared to other regions, it is recommended to dismiss Region 2 for exploitation. Instead, it is advised to focus on Region 1, followed by Region 0, as they present lower risks and better potential profits.

## Conclusion

- The datasets are complete and do not require data completion, data conversion, or removal of null values, making the results highly reliable.
- The linear regression technique provided acceptable validation values for the normal progression of the research.
- The two best exploitation areas with the highest values were estimated, and the region that could lead to losses in the case of eventual exploitation was discarded.